src/modules/base.ts

```diff
@@ -1,119 +1,137 @@
 /**
  * Base module for all API modules
  *
- * @module modules/base
- *
  * @since 1.0.0
  */
 import commsUtils from '../utils/comms';
-import {name as PROTOCOL_AGENT_NAME, version as PROTOCOL_AGENT_VERSION} from '../../package.json';
 
 const PROTOCOL_NAME = 'purecloud-client-apps';
 
-const ACTION_NAME_KEY = 'action';
-const PROTOCOL_NAME_KEY = 'protocol';
-const PROTOCOL_AGENT_NAME_KEY = 'protocolAgentName';
-const PROTOCOL_AGENT_VERSION_KEY = 'protocolAgentVersion';
+interface ProtocolDetails {
+    name: string;
+    agentName: string;
+    agentVersion: string;
+}
+
+export interface SDKMessagePayload {
+    action: string;
+    hook?: string;
+    protocol: string;
+    protocolAgentName: string;
+    protocolAgentVersion: string;
+}
+
+export type MessageListener = (event: SDKMessagePayload) => void;
+export type MessagePayloadFilter = (payload: SDKMessagePayload) => boolean;
+
+type ListenerOptions = Partial<{
+    /** 
+     * Set to true to only invoke this listener once; it will be removed after first invocation.
+     * false by default.  null/undefined will use default.
+     */
+    once: boolean;
+    /**
+     * Provide a function to further filter the invocation of the listener;
+     * The listener will be called if this method returns a truthy value. null by default.  undefined will also use default.
+     */
+    msgPayloadFilter: MessagePayloadFilter | null;
+}>
+
+interface ListenerConfig {
+    listener: MessageListener;
+    options: ListenerOptions;
+}
 
 /**
  * Base Class for PureCloud Client App APIs
  *
  * @since 1.0.0
  */
 class BaseApi {
+    private _targetPcOrigin: string;
+    private _protocolDetails: ProtocolDetails;
+
+    // ----- Message Listening
```

Comment:
Tested lifecycle, once, window message listener cleanup and reactivation when the registered listener count goes to zero and then starts re-adding listeners.  All is looking good 👍